07_VirtualFileSystem/02_VirtualFileSystem.md

```diff
@@ -0,0 +1,352 @@
+* A target folder (that is the folder where the file system will be accessible by the OS) 
+* The target device (in our simple scenario this parameter is going to be mostly ignored since the os will not support any i/o device)
+
+There can be others of course configuration parameters like access permission, driver configuration attributes, etc. For now we haven't implemented a file system yet (we will do soon), but let's assume that our os has a driver for the `USTAR` fs (the one we will implement later), and that the following functions are already implemented: 
+
+```c
+int ustar_open(char *path, int flags);
+int ustar_close(int ustar_fd);
+void ustar_read(int ustar_fd, void *buffer, size_t count);
+int ustar_close(int ustar_fd)
+``` 
+
+For the mount and umount operation we will need two functions: 
+
+* The first one for mounting, let's call it for example `vfs_mount`, to work it will need at least the parameters explained above: 
+
+```c
+int vfs_mount(char *device, char *target, char *fs_type);
+```
+
+Once called it will simply add a new item in the mountpoint on the first available position, and will populate the data mountpoint data structure with the information provided, for example using the array approach, if a free spot is found at index `i` to add a new file sytem we will have something like: 
+
+```c
+mountpoints[i].device = device;
+mountpoints[i].tpye = type;
+mountpoints[i].mountpoint = target;
+mountpoints[i[.operations = NULL 
+```
+
+the last line will be populated soon, for now let's leave it to null. 
+
+* The second instruction is for umounting, in this case since we are just unloading the file device from the system, we don't need to know what type is it, so technically we need either the target device or the target folder, the function can actually accept both parameters, but use only one of them, let's call it `vfs_umount`: 
+
+```c
+int vfs_umount(char *device, char *targe);
+```
+
+In this case we need to find the item in the list that contains the required file system, and if found remove it from the list/tree. In our case since we are using an array we need to clear all the fields in the array at the position containing our fs. 
+
+One thing that we should keep in mind is that using an array, once we umount a file system we are just marking that position as available again, so this means that it can be in the middle of the array, and there can be mounted file system after, so we need to keep in mind this while searching for the next free mouuntpoint.
+
+#### A short diversion: the initialization
+
+When the kernel first boot up, of course there is no file system mounted, and all data structures are not allocated, if we decide to use a linked list for example, before the initialization the pointer will point to nothing (or better to garbage) so we will need to allocate the first item of the list to have the first fs accessible. 
```

Comment:
the phrasing could be improved: "and not all data structures are allocated" is more natural I think.